packages/apps/job-launcher/server/src/modules/job/routing-protocol.service.ts

```diff
@@ -1,34 +1,249 @@
+      this.reputationOracles.map((_, i) => i),
+    );
+
+    this.oracleOrder = this.chains.reduce((acc: OracleOrder, chainId) => {
+      acc[chainId] = this.reputationOracles.reduce(
+        (oracleAcc, reputationOracle) => {
+          oracleAcc[reputationOracle] = {
+            [Role.ExchangeOracle]: {},
+            [Role.RecordingOracle]: {},
+          };
+          return oracleAcc;
+        },
+        {} as {
+          [reputationOracle: string]: {
+            [oracleType: string]: { [jobType: string]: string[] };
+          };
+        },
+      );
+      return acc;
+    }, {} as OracleOrder);
+
+    this.oracleIndexes = this.chains.reduce((acc: OracleIndex, chainId) => {
+      acc[chainId] = this.reputationOracles.reduce(
+        (oracleAcc, reputationOracle) => {
+          oracleAcc[reputationOracle] = {
+            [Role.ExchangeOracle]: {},
+            [Role.RecordingOracle]: {},
+          };
+          return oracleAcc;
+        },
+        {} as {
+          [reputationOracle: string]: {
+            [oracleType: string]: { [jobType: string]: number };
+          };
+        },
+      );
+      return acc;
+    }, {} as OracleIndex);
   }
 
-  private generatePriorityOrder() {
-    const priorityOrder = Array.from(Array(this.chains.length).keys());
-    for (let i = priorityOrder.length - 1; i > 0; i--) {
-      const j = Math.floor(Math.random() * (i + 1));
-      [priorityOrder[i], priorityOrder[j]] = [
-        priorityOrder[j],
-        priorityOrder[i],
-      ];
-    }
-    return priorityOrder;
+  private shuffleArray<T>(array: T[]): T[] {
+    return array.sort(() => Math.random() - 0.5);
   }
 
   public selectNetwork(): ChainId {
-    const chainId = this.chains[this.priorityOrder[this.currentIndex]];
-    this.currentIndex = (this.currentIndex + 1) % this.chains.length;
+    const chainId =
+      this.chains[this.chainPriorityOrder[this.chainCurrentIndex]];
+    this.chainCurrentIndex = (this.chainCurrentIndex + 1) % this.chains.length;
     return chainId;
   }
+
+  public selectReputationOracle(): string {
+    const reputationOracle =
+      this.reputationOracles[
+        this.reputationOraclePriorityOrder[this.reputationOracleIndex]
+      ];
+
+    this.reputationOracleIndex =
+      (this.reputationOracleIndex + 1) % this.reputationOracles.length;
+    return reputationOracle;
+  }
+
+  public selectOracleFromAvailable(
+    availableOracles: any[],
+    oracleType: string,
+    chainId: ChainId,
+    reputationOracle: string,
+    jobType: string,
+  ): string {
+    const oraclesOfType = availableOracles
+      .filter((oracle) => oracle.role === oracleType)
+      .map((oracle) => oracle.address);
+
+    if (!oraclesOfType.length) return '';
+
+    const latestOraclesHash = hashString(JSON.stringify(availableOracles));
+
+    if (
+      !this.oracleOrder[chainId][reputationOracle][oracleType][jobType] ||
+      this.oraclesHash !== latestOraclesHash
```

Comment:
Each set of [chainId][reputationOracle][oracleType][jobType] should have a different hash, otherwise you will be shuffling everytime